# [Complementing a Strand of DNA](https://rosalind.info/problems/revc/)

In [79]:
def reverse_complement_naive(dna_string):
    if not dna_string: 
        return "Invalid string."
    dna_string = dna_string[::-1]
    for i in range(len(dna_string)):
        #inefficient since new strings are created multiple times in loop
        if dna_string[i] == 'A':
            dna_string = dna_string[:i] + 'T' + dna_string[i+1::]
        elif dna_string[i] == 'T':
            dna_string = dna_string[:i] + 'A' + dna_string[i+1::]
        elif dna_string[i] == 'C':
            dna_string = dna_string[:i] + 'G' + dna_string[i+1::]
        elif dna_string[i] == 'G':
            dna_string = dna_string[:i] + 'C' + dna_string[i+1::]
    return dna_string 

In [64]:
def reverse_complement_list_comprehension(dna_string):
    if not dna_string: 
        return "Invalid string."
    dna_string = dna_string[::-1]
    complement_dict = {'A':'T', 'T':'A', 'C':'G', 'G': 'C'}
    complemented = ''.join([complement_dict.get(base, base) for base in dna_string])
    return complemented

In [65]:
def reverse_complement_generator(dna_string):
    if not dna_string:
        return "Invalid string."
    complementDict = {'A':'T', 'T':'A', 'C':'G', 'G': 'C'}
    return ''.join(complementDict.get(base, base) for base in reversed(dna_string))

In [66]:
from utils import dna_utils

In [67]:
dna_string = dna_utils.read_dna_from_txt_file("../data/rosalind_revc.txt")

In [68]:
print(dna_utils.benchmark(reverse_complement_naive, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_naive, dna_string))
print(dna_utils.benchmark(reverse_complement_list_comprehension, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_list_comprehension, dna_string))
print(dna_utils.benchmark(reverse_complement_generator, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_generator, dna_string))

function reverse_complement_naive: time: 1.1339187622070312, dna_length: 909
('Total Memory Usage: 1091.0KB', 'Peak Memory Usage: 3.958984375KB')
function reverse_complement_list_comprehension: time: 0.07605552673339844, dna_length: 909
('Total Memory Usage: 56.0KB', 'Peak Memory Usage: 9.50390625KB')
function reverse_complement_generator: time: 0.12421607971191406, dna_length: 909
('Total Memory Usage: 56.0KB', 'Peak Memory Usage: 8.833984375KB')


In [72]:
dna_string = dna_utils.create_dna_string(10000)
print(dna_utils.benchmark(reverse_complement_naive, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_naive, dna_string))
print(dna_utils.benchmark(reverse_complement_list_comprehension, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_list_comprehension, dna_string))
print(dna_utils.benchmark(reverse_complement_generator, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_generator, dna_string))

function reverse_complement_naive: time: 18.25118064880371, dna_length: 10000
('Total Memory Usage: 56.0KB', 'Peak Memory Usage: 29.58203125KB')
function reverse_complement_list_comprehension: time: 0.6330013275146484, dna_length: 10000
('Total Memory Usage: 112.0KB', 'Peak Memory Usage: 102.845703125KB')
function reverse_complement_generator: time: 1.3058185577392578, dna_length: 10000
('Total Memory Usage: 160.0KB', 'Peak Memory Usage: 93.3447265625KB')


In [77]:
dna_string = dna_utils.create_dna_string(100000)
print(dna_utils.benchmark(reverse_complement_naive, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_naive, dna_string))
print(dna_utils.benchmark(reverse_complement_list_comprehension, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_list_comprehension, dna_string))
print(dna_utils.benchmark(reverse_complement_generator, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_generator, dna_string))

function reverse_complement_naive: time: 1182.168960571289, dna_length: 100000
('Total Memory Usage: 1222.0KB', 'Peak Memory Usage: 296.21484375KB')
function reverse_complement_list_comprehension: time: 6.625890731811523, dna_length: 100000
('Total Memory Usage: 112.0KB', 'Peak Memory Usage: 977.658203125KB')
function reverse_complement_generator: time: 10.465145111083984, dna_length: 100000
('Total Memory Usage: 160.0KB', 'Peak Memory Usage: 880.2666015625KB')


In [78]:
dna_string = dna_utils.create_dna_string(1000000)
print(dna_utils.benchmark(reverse_complement_naive, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_naive, dna_string))
print(dna_utils.benchmark(reverse_complement_list_comprehension, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_list_comprehension, dna_string))
print(dna_utils.benchmark(reverse_complement_generator, dna_string))
print(dna_utils.benchmark_memory(reverse_complement_generator, dna_string))

function reverse_complement_naive: time: 108399.17182922363, dna_length: 1000000
('Total Memory Usage: 2663.0KB', 'Peak Memory Usage: 2932.9375KB')
function reverse_complement_list_comprehension: time: 44.48819160461426, dna_length: 1000000
('Total Memory Usage: 112.0KB', 'Peak Memory Usage: 10203.970703125KB')
function reverse_complement_generator: time: 67.87610054016113, dna_length: 1000000
('Total Memory Usage: 854.0KB', 'Peak Memory Usage: 9228.3505859375KB')


## Benchmark Results:
## 1. Rosalind test DNA string (854 bases)
    function reverse_complement_naive: time: 1.1339187622070312, dna_length: 909
    ('Current Memory Usage: 1091.0KB', 'Peak Memory Usage: 3.958984375KB')
    function reverse_complement_list_comprehension: time: 0.07605552673339844, dna_length: 909
    ('Current Memory Usage: 56.0KB', 'Peak Memory Usage: 9.50390625KB')
    function reverse_complement_generator: time: 0.12421607971191406, dna_length: 909
    ('Current Memory Usage: 56.0KB', 'Peak Memory Usage: 8.833984375KB')

## 2. Generated DNA string (10,000 bases)
    function reverse_complement_naive: time: 18.25118064880371, dna_length: 10000
    ('Current Memory Usage: 56.0KB', 'Peak Memory Usage: 29.58203125KB')
    function reverse_complement_list_comprehension: time: 0.6330013275146484, dna_length: 10000
    ('Current Memory Usage: 112.0KB', 'Peak Memory Usage: 102.845703125KB')
    function reverse_complement_generator: time: 1.3058185577392578, dna_length: 10000
    ('Current Memory Usage: 160.0KB', 'Peak Memory Usage: 93.3447265625KB')

## 3. Generated DNA string (100,000 bases)
    function reverse_complement_naive: time: 1182.168960571289, dna_length: 100000
    ('Current Memory Usage: 1222.0KB', 'Peak Memory Usage: 296.21484375KB')
    function reverse_complement_list_comprehension: time: 6.625890731811523, dna_length: 100000
    ('Current Memory Usage: 112.0KB', 'Peak Memory Usage: 977.658203125KB')
    function reverse_complement_generator: time: 10.465145111083984, dna_length: 100000
    ('Current Memory Usage: 160.0KB', 'Peak Memory Usage: 880.2666015625KB')


## 4. Generated DNA string (1,000,000 bases)   
    function reverse_complement_naive: time: 108399.17182922363, dna_length: 1000000
    ('Current Memory Usage: 2663.0KB', 'Peak Memory Usage: 2932.9375KB')
    function reverse_complement_list_comprehension: time: 44.48819160461426, dna_length: 1000000
    ('Current Memory Usage: 112.0KB', 'Peak Memory Usage: 10203.970703125KB')
    function reverse_complement_generator: time: 67.87610054016113, dna_length: 1000000
    ('Current Memory Usage: 854.0KB', 'Peak Memory Usage: 9228.3505859375KB')
    


## Result:
### Time: 
reverse_complement_list_comprehension < reverse_complement_generator << reverse_complement_naive
### Memory Retention: 
reverse_complement_list_comprehension < reverse_complement_generator << reverse_complement_naive
### Peak Memory: 
reverse_complement_naive << reverse_complement_generator < reverse_complement_list_comprehension  
### Reasoning: 
Naive method is the worst due to several reasons:
- Python for loop is generally slower than list comprehension, because list comprehension implementation is optimized under the hood
- ''.join is much more performant than creating a new string in each for loop since join is implemented in C, and strings are immutable in Python. String slicing is also generally O(n) in Python.

When comparing the list comprehension method vs the generator method, it is important to consider the execution steps:
- List comprehension: 
    1. List comprehension executes, and the list is created and stored in memory
    2. The Join operator iterates over each element in the list and adds it to the running string
    
- Generators:
    1. Generator function is created
    2. Join operator requests each item from the generator 
    3. For each request, the generator executes the next iteration and returns the value
    4. Join add this to the running string, and moves on to the next item returned by the generator until it is exhausted

The generative method may appear to be slower because the generator runs the iterator function each time, but its peak memory usage is much lower which can be especially useful for large DNA strings. For context, a DNA string is on average ~3 billion bases long. However, memory retention for the generator operation at 1,000,000 bases is relatively higher showing that it may benefit from better garbage collection.  
